### PR TITLE
move code that can be shared between clients to common

### PR DESCRIPTION
**Pull request by moolitayer**
This pr's main change is to move as much code as possible to common.rb so code can be shared between client as originally designed. A mixin is introduces to share code between common and clients.

### CODE REVIEW

**Comment by abonas** [comment]
it's unclear to me what you are trying to do here.
I suggest as first step:
a. make a clear PR title that explains the purpose of the PR
b. squash the 5 commits to fewer, plus one of them is named wip
c. check why tests fail
cc @simon3z

**Comment by moolitayer** [comment]
@abonas I've applied to your comments.

**Comment by abonas** [comment]
@simon3z please review. if you think it's good to go you can merge and release.
I still don't understand why all those changes are required including the several renames to modules and parameters, and why most of the client need to be refactored, but perhaps you discussed this with @moolitayer and you have a better insight than me.

**Comment by jimmidyson** [comment]
Does the change usage for existing users?

File: lib/kubeclient.rb [hunk 1]
```diff
@@ -36,43 +38,18 @@ def initialize(uri,
...
-      if auth_options[:user]
-        @basic_auth_user = auth_options[:user]
-        @basic_auth_password = auth_options[:password]
-      elsif auth_options[:bearer_token]
-        bearer_token(auth_options[:bearer_token])
-      elsif auth_options[:bearer_token_file]
-        validate_bearer_token_file(auth_options[:bearer_token_file])
-        bearer_token(File.read(auth_options[:bearer_token_file]))
-      end
+      init_mixin(uri, '/api', version, ssl_options: ssl_options, auth_options: auth_options)
```
Comment (simon3z) [inline lib/kubeclient.rb]: @moolitayer rename to `initialize_client`

**Comment by abonas** [comment]
> Does the change usage for existing users?

It looks like it does. for example auth option parameter 'user' was renamed
https://github.com/abonas/kubeclient/pull/114/files#diff-04c6e90faac2675aa89e2176d2eec7d8R75

**Comment by moolitayer** [comment]
@jimmidyson yes: auth_options.name => auth_options.username

**Comment by jimmidyson** [comment]
@moolitayer Not that it affects me but why the change? Seems redundant & would break stuff for some users (although I bet there's not many actually using basic auth).

**Comment by simon3z** [comment]
All patches are easy to review singularly (except one, because of the new indentation)

- `rename auth_option s/user/username/` :+1:
- `list 'auth_options' keys in ctor` :+1:
- `store auth_options as hash instead of attributes` :+1:
- `move all common code to common.rb` :-1: a couple of comments (it actually touches few things, it's just diff artifacts)
- `changes in validate_auth_options` :+1:
- `add attr accessors` :+1:

cc @abonas @jimmidyson

**Comment by simon3z** [comment]
> @moolitayer Not that it affects me but why the change? Seems redundant & would break stuff for some users (although I bet there's not many actually using basic auth).

@jimmidyson at the moment we kept breaking compatibility (e.g. when we moved the `ssl_options` to the constructor)
I acutally like the rename because it makes the options more consistent `username/password` vs. `user/password`.
For us it's better to change it now if possible.

Maybe @moolitayer can introduce a one-liner compatibility layer:

    @auth_options['username'] = @auth_options['user'] if @auth_options['user']

**Comment by abonas** [comment]
> @jimmidyson at the moment we kept breaking compatibility (e.g. when we moved the ssl_options to the constructor)
> I acutally like the rename because it makes the options more consistent username/password vs. user/password.
> For us it's better to change it now if possible.
> 
> Maybe @moolitayer can introduce a one-liner compatibility layer:
> 
> @auth_options['username'] = @auth_options['user'] if @auth_options['user']

:+1:

**Comment by moolitayer** [comment]
maintaining backward compatibility and test it. comment about it in commit.
expand commit message of add attr accessors

**Comment by simon3z** [comment]
LGTM :+1: (despite the apparent large change in `move all common code to common.rb` which is just indentation diff artifacts). Let's queue this behind #115

File: lib/kubeclient/common.rb [hunk 2]
```diff
@@ -1,255 +1,299 @@
...
+        result_hash[key_name] = send(method_name)
       end
+    end
 
-      def api
-        response = handle_exception do
-          create_rest_client.get(@headers)
-        end
-        JSON.parse(response)
+    def resource_name(entity_type)
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @moolitayer you mistakenly inverted the order of the two methods `resource_name` and `proxy_url`

`resource_name` was after `proxy_url`

**Comment by moolitayer** [comment]
fix inverted methods

File: lib/kubeclient/common.rb [hunk 3]
```diff
@@ -1,255 +1,299 @@
...
+
+    def initialize_client(
+      uri,
+      path,
+      version = nil,
+      ssl_options: {
+        client_cert: nil,
+        client_key: nil,
+        ca_file: nil,
+        verify_ssl: nil
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @moolitayer I think it's safer to set `verify_ssl` to `OpenSSL::SSL::VERIFY_PEER` here as well.

**Comment by simon3z** [comment]
Other than the missing `OpenSSL::SSL::VERIFY_PEER` in `common.rb`, the rest LGTM :+1: ready for merge.

**Comment by moolitayer** [comment]
@abonas @simon3z

**Comment by simon3z** [comment]
@abonas LGTM :+1: ready for merge

**Pull request merged**